Support xspi on rt700

SuperHeroAbner opened the pull request:
Add Zephyr XSPI driver as memc driver.
Add xspi psram and flash driver to handle those different memory cases. The goal is using them to cover all PSRAM and FLASH types. Currently they can only work with the memory on NXP RT700 board.
Porting them to the samples drivers/memc and drivers/spi_flash. RT700 cpu0 supports XSPI0(Flash used in XIP), XSPI1(PSRAM as default), XSPI2(PSRAM). The cpu2 supports XSPI2(PSRAM).

SuperHeroAbner committed:
dts: Enable the xspi flash and psram cases on NXP RT700

Move nxp,xspi.yaml from spi bindings to mtd, it should be flash
controllor rather than simple bus controller.
Delete the xspi flash mx25um51345g bindling yaml file which is useless.
Enable flash and two psrams on RT700 core0 and one psram on core1.

Signed-off-by: Ruijia Wang <ruijia.wang@nxp.com>

SuperHeroAbner committed:
boards: Update related configuration for RT700 xspi case

Add flash and psram dts based on RT700 board resource.
Add the clock and power setting for XSPI in board specfic initilization
function.

Signed-off-by: Ruijia Wang <ruijia.wang@nxp.com>

hakehuang commented:
> @hakehuang Can you help to trigger the unit test?

do you want to enable all spi flash testing  based on this? if so, please update the board support list.

hakehuang commented on samples/drivers/spi_flash/src/main.c:
```diff
@@ -56,6 +56,8 @@
 #define SPI_FLASH_COMPAT nordic_qspi_nor
 #elif DT_HAS_COMPAT_STATUS_OKAY(renesas_ra_ospi_b_nor)
 #define SPI_FLASH_COMPAT renesas_ra_ospi_b_nor
+#elif DT_HAS_COMPAT_STATUS_OKAY(nxp_xspi_flash)
```
in this sample, you need change the sample.yaml as below
```
  sample.drivers.spi.flash:
    filter: dt_compat_enabled("jedec,spi-nor") or dt_compat_enabled("st,stm32-qspi-nor")
      or dt_compat_enabled("st,stm32-ospi-nor") or dt_compat_enabled("st,stm32-xspi-nor")
      or dt_compat_enabled("nordic,qspi-nor") or dt_compat_enabled("jedec,mspi-nor")
      or dt_compat_enabled("nxp,xspi-nor")
```

SuperHeroAbner replied:
Added the "nxp,xspi-flash". Does the name have to be the same as the suffix nor? I'm plan to use this driver for all flash support in the future. Not sure the road map of this part in Zephyr.

SuperHeroAbner commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cm33_cpu0.dts:
```diff
@@ -26,6 +26,7 @@
 		i2s-tx = &sai0;
 		sdhc0 = &usdhc0;
 		rtc = &rtc0;
+		psram0 = &psram0;
```
Use sram-ext

JarmouniA reviewed (changes requested):
Samples changes should be in a separate commit, and should be last. Also, Samples are not actual tests, the driver should be tested with `tests/drivers/memc/ram` and `tests/drivers/flash`.

hakehuang commented:
@SuperHeroAbner , with

samples/drivers/memc/sample.drivers.memc | v4.2.0-2105-g235acffa57db | failed | Timeout | mimxrt700_evk/mimxrt798s/cm33_cpu0 | samples/drivers/memc
-- | -- | -- | -- | -- | --

```
*** Booting Zephyr OS build v4.2.0-2105-g235acffa57db ***
Writing to memory region with base 0x8000000, size 0x400000

```

SuperHeroAbner commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_cm33_cpu0.yaml:
```diff
@@ -24,4 +24,5 @@ supported:
   - usb_device
   - watchdog
   - hwinfo
+  - flash
```
@hakehuang Please check the JP45, should connect 1-2. Shall I add this in certain place in doc?

zejiang0jason commented on drivers/flash/flash_mcux_xspi.c:
```diff
@@ -0,0 +1,490 @@
+
+	ret = flash_mcux_xspi_write_enable(dev, 0, true);
+	if (ret < 0) {
+		return ret;
```
this may return with irq locked, should unlock irq if `irq_lock` is called in this function. Please also check other `return` code in this function.

zejiang0jason commented on drivers/flash/flash_mcux_xspi.c:
```diff
@@ -0,0 +1,490 @@
+{
+	struct flash_mcux_xspi_data *devData = dev->data;
+	const struct device *xspi_dev = devData->xspi_dev;
+	xspi_transfer_t flashXfer;
```
should init this value like `xspi_transfer_t flashXfer = [0}`?

zejiang0jason commented on drivers/flash/flash_mcux_xspi.c:
```diff
@@ -0,0 +1,490 @@
+	uint32_t key = 0;
+	int ret;
+
+	if (offset % SPI_NOR_SECTOR_SIZE) {
```
use `if (0 != (offset % SPI_NOR_SECTOR_SIZE))`, should use a `bool` type.

zejiang0jason commented on drivers/flash/flash_mcux_xspi.c:
```diff
@@ -0,0 +1,490 @@
+{
+	struct flash_mcux_xspi_data *data = dev->data;
+	const struct device *xspi_dev = data->xspi_dev;
+	uint32_t value = 0x02;
```
can we use some macro or some definition? I guess this value is a standard value. Using meaningful definition is better.

zejiang0jason commented on drivers/flash/flash_mcux_xspi.c:
```diff
@@ -0,0 +1,490 @@
+	for (int i = 0; i < size / SPI_NOR_SECTOR_SIZE; i++) {
+		ret = flash_mcux_xspi_write_enable(dev, 0, true);
+		if (ret < 0) {
+			return ret;
```
there are similar case in other functions.

JarmouniA reviewed (dismissed):
Move Samples changes to separate commit, introduce binding with or before the driver, introduce each driver in a separate commit.

SuperHeroAbner committed:
drivers: xspi: add NXP xspi driver

Add mcux xspi driver suppport. Add the flash and psram driver support
based on xspi driver.

Signed-off-by: Ruijia Wang <ruijia.wang@nxp.com>

SuperHeroAbner committed:
soc: Update code to get flash size from dts

Update the Kconfig to parse the dts.

Signed-off-by: Ruijia Wang <ruijia.wang@nxp.com>

SuperHeroAbner committed:
samples: Update XSPI memc and flash samples on RT700

Add memc sample for XSPI PSRAM.
Add spi_flash sample for XSPI Nor Flash.

Signed-off-by: Ruijia Wang <ruijia.wang@nxp.com>

SuperHeroAbner commented:
@hakehuang I linked one hal_nxp commit, please test them together again. @zejiang0jason @Albort12138 Please help to review this patch for hal_nxp. If it's real issue, the cache64/xcache driver in SDK also takes it. https://github.com/zephyrproject-rtos/hal_nxp/pull/597

hakehuang commented:
> @hakehuang I linked one hal_nxp commit, please test them together again

test PASS on v4.2.0-2108-ge629ecd05e10.

Albort12138 commented on drivers/memc/memc_mcux_xspi.c:
```diff
@@ -0,0 +1,188 @@
+		return ret;
+	}
+
+	if (config.ptrAhbAccessConfig != NULL) {
```
No other configurations to set ptrAhbAccessConfig as NULL, this if condition can be removed

Albort12138 commented on drivers/memc/memc_mcux_xspi.c:
```diff
@@ -0,0 +1,188 @@
+			}, \
+			.ptrIpAccessConfig = &(xspi_ip_access_config_t){ \
+				.ipAccessTimeoutValue = 0xFFFFFFFF, \
+				.ptrSfpFradConfig = NULL, \
```
Should allow developers to set SFP configurations

JarmouniA reviewed (commented):
I can't seem to be able to dismiss my change request, so If someone can, please do.

decsny commented:
> I can't seem to be able to dismiss my change request, so If someone can, please do.

It's not showing up for me either , maybe it's a bug. Perhaps the author fixing the PR conflict can help

dleach02 commented:
@SuperHeroAbner please address the merge conflicts

hakehuang commented:
board testing pass. on v4.2.0-5922-ga36ad82ca689

```
*** Booting Zephyr OS build v4.2.0-5922-ga36ad82ca689 ***

flash-controller@0 SPI flash testing
==========================

Perform test on single sector
Test 1: Flash erase
Flash erase succeeded!

Test 2: Flash write
Attempting to write 4 bytes
Data read matches data written. Good!!

```

SuperHeroAbner commented:
@dleach02 Can we merge this PR for 4.3.0 release together with https://github.com/zephyrproject-rtos/hal_nxp/pull/597
